pkg/operators/ebpf/stats.go

```diff
@@ -0,0 +1,527 @@
+			DefaultValue: "false",
+			TypeHint:     api.TypeBool,
+			Title:        "Per-program statistics",
+		},
+	}
+}
+
+func (o *ebpfOperator) Priority() int {
+	return 0
+}
+
+type ebpfOperatorDataInstance struct {
+	bpfOperator *ebpfOperator
+	ds          datasource.DataSource
+	interval    time.Duration
+	count       int
+	done        chan struct{}
+
+	// used to emit incremental values
+	oldProgStats map[ebpf.ProgramID]progStat
+
+	// if true stats are collected with programs granularity
+	programStats bool
+
+	nodeNameField datasource.FieldAccessor
+	nodeName      string
+
+	// gadget-specific fields
+	gadgetIDField    datasource.FieldAccessor
+	gadgetNameField  datasource.FieldAccessor
+	gadgetImageField datasource.FieldAccessor
+
+	// low-level fields
+	progIDField   datasource.FieldAccessor
+	progNameField datasource.FieldAccessor
+
+	// stats fields
+	runtimeField   datasource.FieldAccessor
+	runcountField  datasource.FieldAccessor
+	mapMemoryField datasource.FieldAccessor
+	mapCountField  datasource.FieldAccessor
+}
+
+func (i *ebpfOperatorDataInstance) Name() string {
+	return "ebpfdataoperator"
+}
+
+type stat struct {
+	gadgetID    string
+	gadgetName  string
+	gadgetImage string
+
+	programID   uint32
+	programName string
+
+	runtime   uint64
+	runcount  uint64
+	mapMemory uint64
+	mapCount  uint64
+}
+
+type progStat struct {
+	runtime  uint64
+	runcount uint64
+}
+
+func (i *ebpfOperatorDataInstance) sendStats(stats []stat) error {
+	arr, err := i.ds.NewPacketArray()
+	if err != nil {
+		return fmt.Errorf("creating new packet: %w", err)
+	}
+
+	for _, stat := range stats {
+		d := arr.New()
+
+		if i.nodeNameField != nil {
+			i.nodeNameField.PutString(d, i.nodeName)
+		}
+		i.gadgetIDField.PutString(d, stat.gadgetID)
+		i.gadgetNameField.PutString(d, stat.gadgetName)
+		i.gadgetImageField.PutString(d, stat.gadgetImage)
+		if i.programStats {
+			i.progIDField.PutUint32(d, stat.programID)
+			i.progNameField.PutString(d, stat.programName)
+		}
+		i.runtimeField.PutUint64(d, stat.runtime)
+		i.runcountField.PutUint64(d, stat.runcount)
+		i.mapMemoryField.PutUint64(d, stat.mapMemory)
+		i.mapCountField.PutUint64(d, stat.mapCount)
+
+		arr.Append(d)
+	}
+
+	return i.ds.EmitAndRelease(arr)
+}
+
+func getProgramStats(cache map[ebpf.ProgramID]progStat, id ebpf.ProgramID) (progStat, error) {
+	if stat, ok := cache[id]; ok {
+		return stat, nil
+	}
+
+	prog, err := ebpf.NewProgramFromID(id)
+	if err != nil {
+		return progStat{}, err
+	}
+	defer prog.Close()
+
+	pi, err := prog.Info()
+	if err != nil {
+		return progStat{}, err
+	}
+
+	runtime, _ := pi.Runtime()
+	runcount, _ := pi.RunCount()
+
+	return progStat{
+		runtime:  uint64(runtime),
+		runcount: runcount,
+	}, nil
+}
+
+func (i *ebpfOperatorDataInstance) getStats() ([]stat, error) {
+	stats := make([]stat, 0)
+
+	mapSizes, err := bpfstats.GetMapsMemUsage()
+	if err != nil {
+		return nil, fmt.Errorf("getting map memory usage: %w", err)
+	}
+
+	// TODO: reduce lock contention
```

Comment:
Is it necessary? `getStats` is called from a Ticker, so no parallel calls. And other usages of `mu.Lock()` are not frequent.

Reply:
You're right. I was a bit bothered at the beginning because it was covering a much bigger region, but I think it's fine now.